src/Microsoft.NetCore.Analyzers/Core/Security/DoNotOverloadDeserializationBinderWithoutThrowingAnException.cs

```diff
@@ -0,0 +1,158 @@
+    public class DoNotOverloadSerializationBinderWithoutThrowingAnException : DiagnosticAnalyzer
+    {
+        protected string OverloadingClassMetadataName { get; }
+        enum CacheState
+        {
+            Absent,
+            InProgress,
+            PresentTrue,
+            PresentFalse
+        };
+
+        internal static DiagnosticDescriptor DoNotOverloadSerializationBinderWithoutThrowingAnExceptionRule = SecurityHelpers.CreateDiagnosticDescriptor(
+            "CA2340",
+            nameof(MicrosoftNetCoreAnalyzersResources.DoNotOverloadSerializationBinderWithoutThrowingAnExceptionTitle),
+            nameof(MicrosoftNetCoreAnalyzersResources.DoNotOverloadSerializationBinderWithoutThrowingAnExceptionMessage),
+            DiagnosticHelpers.EnabledByDefaultIfNotBuildingVSIX,
+            helpLinkUri: null,
+            customTags: WellKnownDiagnosticTags.Telemetry);
+
+        public sealed override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+            ImmutableArray.Create<DiagnosticDescriptor>(
+                DoNotOverloadSerializationBinderWithoutThrowingAnExceptionRule);
+
+        public sealed override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            // Security analyzer - analyze and report diagnostics on generated code.
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+
+            context.RegisterCompilationStartAction(
+                (CompilationStartAnalysisContext compilationStartAnalysisContext) =>
+                {
+                    Dictionary<IMethodSymbol, CacheState> methodCache = new Dictionary<IMethodSymbol, CacheState>();
+                    WellKnownTypeProvider typeProvider = WellKnownTypeProvider.GetOrCreate(compilationStartAnalysisContext.Compilation);
+                    INamedTypeSymbol SerializationBinder = null;
+                    if (!typeProvider.TryGetTypeByMetadataName("System.Runtime.Serialization.SerializationBinder", out SerializationBinder))
+                    {
+                        System.Diagnostics.Debug.Fail("Unable to find System.Runtime.Serialization.SerializationBinder threw WellKnownTypeProvider");
+                        return;
+                    }
+
+                    compilationStartAnalysisContext.RegisterSymbolAction(
+                        (SymbolAnalysisContext symbolAnalysisContext) =>
+                        {
+                            IMethodSymbol msym = (IMethodSymbol)symbolAnalysisContext.Symbol;
+                            if (msym.Name == "BindToType")
+                            {
+                                if (ITypeSymbolExtensions.Inherits(msym.ContainingType, SerializationBinder))
+                                {
+                                    IBlockOperation msymOps = msym.GetTopmostOperationBlock(compilationStartAnalysisContext.Compilation);
```

Comment:
I looked at your logic and I believe you are trying to flag all methods named BindToType in types deriving from SerializationBinder that don’t throw an exception in its body. Is that correct? If so, a much simpler implementation would be to RegisterOperationBlockStartAction which validates method name and containing type, and if that succeeds registers (1) a nested RegisterOperationAction for OperationKind.Throw, which sets a boolean flag scoped to the OperationBlockStartAction and (2) a nested OperationBlockEndAction that reports a diagnostic if the boolean flag is not set. Let me know if you have any questions on this.

Reply:
This might be a good example: https://github.com/dotnet/roslyn-analyzers/blob/c9d6579f31fd4ad45f3d23187189657343b395a4/src/Microsoft.NetCore.Analyzers/Core/Runtime/DisposeMethodsShouldCallBaseClassDispose.cs#L54-L116

Reply:
That is broadly the original logic, but it has been extended after some research into typical use to also search any function called from BindToType() for throws as well. If any of them throw an exception and are in the same assembly, the test passes. This has to continue functioning for multiple layers of called functions and/or recursive functions. Would there be value in modifying the logic with extra callbacks?

Reply:
Current logic is not performant. You are going from symbol to operations, which may lead to duplicate binding. Instead any analyzer which is interested in semantic analysis of executable code blocks should try to restrict analysis to operations. Additionally, I believe this will simplify the implementation significantly to just a handful of lines - you won't need the dictionary or your recursive local function.

Reply:
To clarify - having additional analyer callbacks is not expensive, as the compiler makes the callback at the exact time it binds/compiles the code unit. However, going from symbols to operations (method bodies) will be expensive as that will create a separate semantic model for the analyzer and lead to duplicate binding for the compiler and analyzer.

Reply:
The problem there is that I have to follow calls to other functions. If I only look at operations, I can see those calls occurring by catching IInvocationOperations, but those don't automatically allow me to see their operations without using their symbols. Since some of those functions may already have been compiled, I can't even identify their symbols and start registering a bunch more callbacks unless doing this will cause the compiler to retroactively call the function that didn't exist at the original time of compilation.

Reply:
@mavasani `DisposeMethodsShouldCallBaseClass` doesn't consider intermediate methods calling `base.Dispose()`, does it?

I guess another approach is to build a call graph for every method invocation, and dictionary of whether a given method throws an exception, and at then at the end, walk through the call graphs of all `BindToType` methods we care about and see if they throw / call another method that throws.

Reply:
> I guess another approach is to build a call graph for every method invocation, and dictionary of whether a given method throws an exception, and at then at the end, walk through the call graphs of all BindToType methods we care about and see if they throw / call another method that throws.

Yes, if you care about following the interprocedural call chain, then you should follow the approach taken by @LLLXXXCCC in her analyzers, like: https://github.com/dotnet/roslyn-analyzers/blob/c9d6579f31fd4ad45f3d23187189657343b395a4/src/Microsoft.NetCore.Analyzers/Core/Security/DoNotCallDangerousMethodsInDeserialization.cs

It would also be good to share the code and move the call graph generator and call graph walker at compilation end from that analyzer into a shared utility.

Reply:
I'm new to this, so perhaps I'm not understanding the performance implications properly, but as I understand it, creating such a graph requires an analysis of every function at every point in any code that is being compiled. This would seem to be very intensive from a compute and memory perspective, when there's no guarantee that they even use SerializationBinders. The current approach only starts the analysis with functions that are eligible for the diagnostic, and builds the graph retroactively. Because of the included cache, the required subset of the graph is created for future reference, but the unvisited nodes are not computed. Wouldn't the current approach be faster?

Reply:
The performance problem from your current implementation comes when you register a symbol action callback and then try to get the top level block operation for that method symbol. At that moment, the block operation that you get is using a different semantic model then what is used by the compiler for binding method bodies, which means that your analyzer will bind/compile all the method bodies one extra time. On the other hand all OperationBlock/Operation callbacks already provide you the entire operation tree with all semantic details, and the operations are generated from the semantic model that was used by the compiler during the actual compile, so it is essentially re-using the work done during regular compilation for semantic analysis.

The rule of thumb is that is if you are planning to do any semantic analysis of executable code blocks/method bodies and/or require to work on operations, you should do so by registering explicit operation callbacks with the compiler, not using our internal API that gets you operation blocks for symbol. That internal API is only meant for interprocedural analysis scenario, where we don't have any option but to do the extra binding work for the invoked method analysis.

We can discuss in person if there is still some confusion.

Reply:
That part makes sense. However, since there is some interprocedural work, how does the impact of rechecking those functions called by BindToType compare to registering a callback for every function in the codebase and then compiling a dictionary of every function? I'd imagine that the simple loop through operations on each function would take some processing power, and the graph would take a lot of memory depending on the size of the code. Meanwhile, this comparatively longer way only checks a few functions. I don't have good assumptions for the ratio between processing times of the two methods of getting operations.